### PR TITLE
ipn,tailfs: updates to support sandboxed MacOS application

### PR DESCRIPTION
**Pull request by oxtoacart**
This includes two commits that together allow the sandboxed Mac application to use security scoped bookmarks to gain and retain access to restricted folders.

### CODE REVIEW

**Review by agottardo — APPROVED** [review]
LGTM

**Comment by bradfitz** [comment]
> updates to support

=> "support"

(verb)

File: tailfs/remote.go [hunk 1]
```diff
@@ -31,6 +31,14 @@ type Share struct {
...
 	// Tailscale GUI".
 	As string `json:"who"`
+
+	// BookmarkData contains security-scoped bookmark data for the Sandboxed
+	// Mac application. The Sandboxed Mac application gains permission to
+	// access the Share's folder as a result of a user selecting it in a file
+	// picker. In order to retain access to it across restarts, it needs to
+	// hold on to a security-scoped bookmark. That bookmark is stored here. See
+	// https://developer.apple.com/documentation/security/app_sandbox/accessing_files_from_the_macos_app_sandbox#4144043
+	BookmarkData []byte `json:"bookmarkData"`
```
Comment (bradfitz) [inline tailfs/remote.go:41]: is this intentionally not omitempty?

File: ipn/ipnlocal/tailfs.go [hunk 2]
```diff
@@ -179,20 +179,20 @@ func (b *LocalBackend) tailfsRemoveShareLocked(name string) (map[string]string,
 	}
 	fs.SetShares(shares)
 
-	return shareNameMap(shares), nil
+	return copyShares(shares), nil
 }
 
-func shareNameMap(sharesByName map[string]*tailfs.Share) map[string]string {
-	sharesMap := make(map[string]string, len(sharesByName))
+func copyShares(sharesByName map[string]*tailfs.Share) map[string]*tailfs.Share {
```
Comment (bradfitz) [inline ipn/ipnlocal/tailfs.go:185]: is this just https://pkg.go.dev/maps#Clone ?

But it converts nil maps to non-nil zero length maps?

Comment (oxtoacart) [reply ipn/ipnlocal/tailfs.go:185]: Oh hey, maps.Clone is nice, will use that.

File: tailfs/remote.go (hunk 1 above)
Comment (oxtoacart) [reply tailfs/remote.go:41]: No, will change.

**Review by bradfitz — COMMENTED** [review]
BTW, do you know about Views and how we use them?

Seems like what you really want is to pass around immutable views instead of clones and only clone on internal mutations. (take existing view, clone into mutable copy, mutate, and save back into a view field on LocalBackend) That's increasingly the pattern we do to prevent runaway clone-defensive garbage (allocation garbage).

**Pull request merged**